include/unifex/find_if.hpp

```diff
@@ -24,63 +25,162 @@
+    auto operator()(const unpack_receiver<Receiver>& /*unused*/, const sequenced_policy&, Iterator begin_it, Iterator end_it, Values&&... values) noexcept {
+      // Sequential implementation
+      return unifex::transform(
+          unifex::just(std::forward<Values>(values)...),
+          [this, begin_it, end_it](auto... values) {
+            for(auto it = begin_it; it != end_it; ++it) {
+              if(std::invoke((Func &&) func_, *it, values...)) {
+                return std::tuple<Iterator, Values...>(it, std::move(values)...);
+              }
+            }
+            return std::tuple<Iterator, Values...>(end_it, std::move(values)...);
+          }
+        );
+    }
+
+    template<typename Iterator, typename... Values>
+    auto operator()(const unpack_receiver<Receiver>& receiver, const parallel_policy&, Iterator begin_it, Iterator end_it, Values&&... values) noexcept {
+      // func_ is safe to run concurrently so let's make use of that
+
+      // NOTE: Assumes random access iterator for now, on the assumption that the policy was accurate
+      constexpr int max_num_chunks = 16;
+      constexpr int min_chunk_size = 8;
+      auto distance = std::distance(begin_it, end_it);
+      auto num_chunks = (distance/max_num_chunks) > min_chunk_size ? max_num_chunks : ((distance+min_chunk_size)/min_chunk_size);
+      auto chunk_size = (distance+num_chunks)/num_chunks;
+
+      // Use bulk_schedule to construct parallelism, but block and use local vector for now
+      return unifex::let(
+        unifex::just(std::vector<Iterator>(num_chunks), std::forward<Values>(values)...),
+        [this, sched = unifex::get_scheduler(receiver), begin_it, chunk_size, end_it, num_chunks](
+            std::vector<Iterator>& perChunkState, Values&... values) {
+          return unifex::transform(
+            unifex::bulk_join(
+              unifex::bulk_transform(
+                unifex::bulk_schedule(std::move(sched), num_chunks),
+                [this, &perChunkState, begin_it, chunk_size, end_it, num_chunks, &values...](std::size_t index){
+                  auto chunk_begin_it = begin_it + (chunk_size*index);
+                  auto chunk_end_it = chunk_begin_it;
+                  if(index < (num_chunks-1)) {
+                    std::advance(chunk_end_it, chunk_size);
+                  } else {
+                    chunk_end_it = end_it;
+                  }
+
+                  for(auto it = chunk_begin_it; it != chunk_end_it; ++it) {
+                    if(std::invoke(func_, *it, values...)) {
+                      perChunkState[index] = it;
+                      return;
+                    }
+                  }
+                  // If not found, return the very end value
+                  perChunkState[index] = end_it;
+                },
+                unifex::par
+              )
+            ),
+            [&perChunkState, end_it, &values...]() mutable -> std::tuple<Iterator, Values...> {
```

Comment:
It's unfortunate you have to pack the results into a tuple just so you can return them from a lambda, only to have to then pipe them through another 'unpack' adapter to get them into multiple set_value() args.

I wonder if this whole thing might be more simply implemented as a custom many_receiver that explicitly implements set_next/set_value/set_error/set_done, etc. rather than using transform() here.

Reply:
Yes, that would certainly be possible. It just would mean more work in custom code while I would expect that an unpack algorithm would be a reasonable generic addition.